src/frame-writer.cpp

```diff
@@ -523,94 +525,216 @@ static enum AVSampleFormat convert_codec_sample_fmt(const AVCodec *codec, std::s
-    if (params.sample_fmt.size() == 0) 
-    {
-        audioCodecCtx->sample_fmt = get_codec_auto_sample_fmt(codec);
-        std::cerr << "Choosing sample format " << av_get_sample_fmt_name(audioCodecCtx->sample_fmt) << " for audio codec " << codec->name << std::endl;
-    } else 
-    {
-        audioCodecCtx->sample_fmt = convert_codec_sample_fmt(codec, params.sample_fmt);
-    }
+        const AVCodec* codec = avcodec_find_encoder_by_name(params.audio_codec.c_str());
+        if (!codec)
+        {
+            std::cerr << "Failed to find the given audio codec: " << params.audio_codec << std::endl;
+            exit(1);
+        }
+
+        AVStream* audioStream = avformat_new_stream(fmtCtx, codec);
+        if (!audioStream)
+        {
+            std::cerr << "Failed to open audio stream" << std::endl;
+            exit(1);
+        }
+
+        AVCodecContext* audioCodecCtx = avcodec_alloc_context3(codec);
+
+        if (params.sample_fmt.empty()) {
+            audioCodecCtx->sample_fmt = get_codec_auto_sample_fmt(codec);
+            std::cerr << "Choosing sample format " << av_get_sample_fmt_name(audioCodecCtx->sample_fmt) << " for audio codec " << codec->name << std::endl;
+        } else {
+            audioCodecCtx->sample_fmt = convert_codec_sample_fmt(codec, params.sample_fmt);
+        }
 #if HAVE_CH_LAYOUT
-    av_channel_layout_from_mask(&audioCodecCtx->ch_layout, get_codec_channel_layout(codec));
+        av_channel_layout_from_mask(&audioCodecCtx->ch_layout, get_codec_channel_layout(codec));
 #else
-    audioCodecCtx->channel_layout = get_codec_channel_layout(codec);
-    audioCodecCtx->channels = av_get_channel_layout_nb_channels(audioCodecCtx->channel_layout);
+        audioCodecCtx->channel_layout = get_codec_channel_layout(codec);
+        audioCodecCtx->channels = av_get_channel_layout_nb_channels(audioCodecCtx->channel_layout);
 #endif
-    audioCodecCtx->sample_rate = params.sample_rate;
-    audioCodecCtx->time_base = (AVRational) { 1, audioCodecCtx->sample_rate };
+        audioCodecCtx->sample_rate = params.sample_rate;
+        audioCodecCtx->time_base = (AVRational) { 1, audioCodecCtx->sample_rate };
 
-    if (fmtCtx->oformat->flags & AVFMT_GLOBALHEADER)
         audioCodecCtx->flags |= AV_CODEC_FLAG_GLOBAL_HEADER;
 
-    int err;
-    if ((err = avcodec_open2(audioCodecCtx, codec, NULL)) < 0)
-    {
-        std::cerr << "(audio) avcodec_open2 failed " << err << std::endl;
-        std::exit(-1);
+        int err;
+        if ((err = avcodec_open2(audioCodecCtx, codec, NULL)) < 0)
+        {
+            std::cerr << "(audio) avcodec_open2 failed " << err << std::endl;
+            exit(1);
+        }
+
+        // Create SwrContext for this stream
+        SwrContext* swrCtx = swr_alloc();
+#if HAVE_CH_LAYOUT
+        AVChannelLayout source_layout = {};
+        av_channel_layout_default(&source_layout, 2);
+
+        av_opt_set_int(swrCtx, "in_sample_rate", audioCodecCtx->sample_rate, 0);
```

Comment:
seems like most of these are duplicated in the two if-else branches, I think we should keep duplication to minimum.